forge-game/src/main/java/forge/game/ability/effects/SeekEffect.java

```diff
@@ -0,0 +1,85 @@
+package forge.game.ability.effects;
+
+import com.google.common.collect.Lists;
+import forge.game.Game;
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.Card;
+import forge.game.card.CardCollection;
+import forge.game.card.CardCollectionView;
+import forge.game.card.CardZoneTable;
+import forge.game.player.Player;
+import forge.game.spellability.SpellAbility;
+import forge.game.trigger.TriggerType;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.Aggregates;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class SeekEffect extends SpellAbilityEffect {
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card source = sa.getHostCard();
+        List<Player> seekers;
+
+        if (sa.hasParam("Defined")) {
+            seekers = AbilityUtils.getDefinedPlayers(source, sa.getParam("Defined"), sa);
+        } else {
+            seekers = Lists.newArrayList(sa.getActivatingPlayer());
+        }
+
+        final Game game = source.getGame();
+        List<String> seekTypes = Lists.newArrayList();
+        if (sa.hasParam("Types")) {
+            seekTypes.addAll(Arrays.asList(sa.getParam("Types").split(",")));
+        } else {
+            seekTypes.add(sa.getParamOrDefault("Type", "Card"));
+        }
+
+        int seekNum = sa.hasParam("Num") ? AbilityUtils.calculateAmount(source, sa.getParam("Num"), sa) : 1;
+
+        for (Player seeker : seekers) {
+            final CardZoneTable triggerList = new CardZoneTable();
+            CardCollectionView lastStateBattlefield = game.copyLastStateBattlefield();
+            CardCollectionView lastStateGraveyard = game.copyLastStateGraveyard();
+            for (String seekType : seekTypes) {
+                CardCollection pool;
+                if (sa.hasParam("DefinedCards")) {
+                    pool = AbilityUtils.getDefinedCards(source, sa.getParam("DefinedCards"), sa);
+                } else {
+                    pool = new CardCollection(seeker.getCardsIn(ZoneType.Library));
+                }
+                pool = (CardCollection)AbilityUtils.filterListByType(pool, seekType, sa);
```

Comment:
i personally hate `filterListByType` and prefer the real `CardLists.getValidCards`

i think filterListByType is from a time where we aren't that flexible and most stuff gets to be hardcoded

Reply:
Sounds good to me – as you can probably tell, I just went through ChangeZoneEffect and copied the needed bits